File: src/XCocoaUtilsPublic/UIKit/src/UIBarButtonItem+XCUP.m
//
//  UIBarButtonItem+XCUP.m
//  XCocoaUtilsPublic
//
//  Created by Deheng Xu on 2020/10/25.
//

#import "UIBarButtonItem+XCUP.h"

CGSize XCUPDefaultBarButtonItemSize = (CGSize){44.0, 44.0};

//@implementation UIBarButtonItem (XCUP)
//
//- (instancetype)initWithXCUPTitle:(NSString *)title target:(id)target selector:(SEL)selector buttonSize:(CGSize)size buttonType:(UIButtonType)type forEvent:(UIControlEvents)event {
//	UIButton *btn = [UIButton buttonWithType:type];
//	CGRect f = {CGPointZero, size};
//	btn.frame = f;
//	[btn setTitle:title forState:UIControlStateNormal];
//	[btn addTarget:target action:selector forControlEvents:event];
//	self = [self initWithCustomView:btn];
//	return self;
//}
//
//- (instancetype)initWithXCUPTitle:(NSString *)title target:(id)target selector:(SEL)selector buttonSize:(CGSize)size buttonType:(UIButtonType)type {
//	return [self initWithXCUPTitle:title target:target selector:selector buttonSize:size buttonType:type forEvent:UIControlEventTouchUpInside];
//}
//
//- (instancetype)initWithXCUPTitle:(NSString *)title target:(id)target selector:(SEL)selector buttonSize:(CGSize)size {
//	return [self initWithXCUPTitle:title target:target selector:selector buttonSize:size buttonType:UIButtonTypeSystem];
//}
//
//- (instancetype)initWithXCUPTitle:(NSString *)title target:(id)target selector:(SEL)selector {
//	return [self initWithXCUPTitle:title target:target selector:selector buttonSize:XCUPDefaultBarButtonItemSize];
//}
//
//- (UIButton *)xcup_customButton {
//	if ([self.customView isKindOfClass:UIButton.class]) {
//		return self.customView;
//	}
//	return nil;
//}
//@end

File: src/XCocoaUtilsPublic/Swift/src/UIBarButtonItem+XCUP.swift
//
//  UIBarButtonItem+XCUP.swift
//  XCocoaUtilsPublic
//
//  Created by Deheng Xu on 2020/10/17.
//

#if os(iOS) //os(iOS)
import UIKit

private protocol ButtonAsCustomView {}

extension UIBarButtonItem: ButtonAsCustomView {


	public convenience init(xcupTitle: String? = nil, titleColor: UIColor = .black, target: AnyObject? = nil, selector: Selector? = nil, buttonType: UIButton.ButtonType = .custom, forEvent: UIControl.Event = .touchUpInside) {
		let btn = UIButton.xcup_button(title: xcupTitle, titleColor: titleColor, target: target, selector: selector, closure: nil, type: buttonType)
		self.init(customView: btn)
	}

	public convenience init(xcupTitle: String? = nil, titleColor: UIColor = .black, closure: ((_ sender: UIButton)->Void)? = nil, buttonType: UIButton.ButtonType = .custom, forEvent: UIControl.Event = .touchUpInside) {
		let btn = UIButton.xcup_button(title: xcupTitle, titleColor: titleColor, target: nil, selector: nil, closure: closure, type: buttonType)
		self.init(customView: btn)
	}

	public func xcup_withCustomViewSize(width: CGFloat = 64.0, height: CGFloat = 44.0) -> Self {
		if let view = self.customView {
			view.frame.size.width = width
			view.frame.size.height = height
		}
		return self
	}

	public func xcup_customButton() -> UIButton? {
		return self.customView as? UIButton
	}

}

#endif
